Make DataBuilder for all project rake task (#783)

Anorwen opened the pull request:
Here is the distilled rake file for issue 783. A few notes about the changes:

- Chromium has an extra tagger called 'FixTagger', will this cause issues if it's used in other projects?
- I altered some of the commands. To build a project individually use 'rails data:build[_name_]'. I pass in parameters to tasks but have not updated the readme until everyone is okay with the change.
- I'm unsure about the best way to execute tasks within a task (the current version seems redundant as invoking one task requires reenabling it before calling it again.

Let me know what you guys think or if you have any questions.

andymeneely commented on lib/tasks/project.rake:
```diff
@@ -0,0 +1,218 @@
+require_all 'lib/taggers/*.rb'
+
+desc 'Holds Git repository information for the project builds.'
+class BuilderData 
```
Let's move this class to its own file in lib/  and then move most of the functionality to that class. Rakefiles are intended to be succinct and we were already straining that convention. Also that would make it more of a real class than the struct that it basically is right now.

Anorwen replied:
I moved the BuilderData class into the lib directory.

andymeneely commented on lib/tasks/all.rake:
```diff
@@ -1,17 +1,21 @@
+    Rake::Task['data:writing'].reenable
+    Rake::Task['data:done']
+    Rake::Task['data:done'].reenable
+  end
```
While this technically probably "works", it's a lot less maintainable because it requires knowledge of how Rakefiles work (like I don't know offhand what `reenable` actually does and I didn't have to look it up before).  I'd rather have a similar format as before and just declare new rake tasks one by one. The builder below just takes in the project parameters and then the tasks stay the same. Actually, all those calls should just go to a `data.rake`  - that makes more sense.

Anorwen replied:
I reverted my changes for invoking rake tasks as dependencies.

andymeneely commented on lib/tasks/project.rake:
```diff
@@ -0,0 +1,218 @@
+require_all 'lib/taggers/*.rb'
+
+desc 'Holds Git repository information for the project builds.'
+class BuilderData 
```
Let's call this `DataBuilder` instead

Anorwen replied:
Changed the name from BuilderData to DataBuilder.

andymeneely commented on lib/tasks/project.rake:
```diff
@@ -0,0 +1,218 @@
+    end
+
+    desc 'Possible Projects include: chromium, httpd, struts, and tomcat.'
+    namespace :project do        
```
I don't think we need another namespace for `project`, I think we can keep the same interface we have with `data:all`, `data:chromium`, etc.

Anorwen replied:
I have pushed a namespace configuration that is identical to the original.

andymeneely commented on lib/tasks/project.rake:
```diff
@@ -0,0 +1,218 @@
+                ErrorOfOmissionTagger,
+                LanguageTagger,
+                LessonsTagger,
+                FixTagger,
```
Per your question - everyone should have a fixtagger, so this is good.

andymeneely commented on lib/tasks/project.rake:
```diff
@@ -0,0 +1,218 @@
+            Rake::Task['data:no_logger'].reenable
+        end
+
+        Rake::Task['data:project:check_exist'].invoke(args[:name])
```
Eh, I think I like the way we had it without the args. I used to have args and decided to go simpler.

Anorwen replied:
I removed all arguments from tasks.

andymeneely reviewed (changes requested):
Excellent! I've got some significant revisions I'm asking for, but that's to be expected with such a big change like this.  In summary I'd like to keep the simplicity of the rake tasks - no command-line parameters, dependencies instead of invoking, and no extra sub-namespaces. Simpler rakefile and move the real building out to a separate `DataBuilder` class. I think that'll strike the balance between being DRY and not being over-parameterized/over-complicated.

Anorwen commented:
I pushed a new version of the rake file to the branch. I've had difficulty testing it because my machine seems to be freezing when interacting with the database (I've waited at least 5 minutes before canceling the tasks). Please let me know if the commands work for you.

I've incorporated the requested changes and think it looks better than before. One small change I'm considering is going back to a rake file convention where the interface for each project is specified. Currently all of the project tasks are declared in 1 file (with 4 DataBuilders). Should I split the data.rake into 4 files, one for each project?

Please let me know of any changes I've missed or other things I should modify.
